### ISSUE TITLE
Integrate bencher for benchmarks

### ISSUE THREAD

**Issue by tusharmath** [tailcallhq/tailcall#1300]
The idea here is to integrate [Bencher] by @epompeii and keep track of performance.

[Bencher]: https://github.com/bencherdev/bencher

It would be great if we could push data of older commits for the last 2-3 months and visualize it on something like — https://bencher.dev/perf/advent-of-code

**Comment by epompeii** [comment]
Thank you for following up @tusharmath!
This is a follow on issue from a discussion on https://github.com/tailcallhq/tailcall/issues/436

I will create a PR that runs on pushes to the `main` branch, following the first example shown here: https://bencher.dev/docs/how-to/github-actions/
A working example of this approach is also available here: https://github.com/bencherdev/example/blob/main/.github/workflows/benchmarks.yml
The perf page for that example is here: https://bencher.dev/perf/example

In order for this to work, you all need to have `BENCHER_API_TOKEN` set as a repository secret.
It will be a bit tedious to add accurate historical data. Though the Bencher API supports backdating reports, getting that historical data would require replaying each past push to `main` in GitHub Actions in chronological order.
With that said, the PR I'll be posting will be forward looking only.

Currently you all have implemented [relative benchmarking](https://bencher.dev/docs/how-to/track-benchmarks/#relative-benchmarking), comparing a single run on main to a single run on a PR here: https://github.com/tailcallhq/tailcall/pull/762
This can also be accomplished with Bencher as well, and this first PR will lay the ground work for you to use [more advanced Threshold](https://bencher.dev/docs/explanation/thresholds/) for detecting performance regressions.

**Comment by alankritdabral** [comment]
> Thank you for following up @tusharmath! This is a follow on issue from a discussion on #436
> 
> I will create a PR that runs on pushes to the `main` branch, following the first example shown here: https://bencher.dev/docs/how-to/github-actions/ A working example of this approach is also available here: https://github.com/bencherdev/example/blob/main/.github/workflows/benchmarks.yml The perf page for that example is here: https://bencher.dev/perf/example
> 
> In order for this to work, you all need to have `BENCHER_API_TOKEN` set as a repository secret. It will be a bit tedious to add accurate historical data. Though the Bencher API supports backdating reports, getting that historical data would require replaying each past push to `main` in GitHub Actions in chronological order. With that said, the PR I'll be posting will be forward looking only.
> 
> Currently you all have implemented [relative benchmarking](https://bencher.dev/docs/how-to/track-benchmarks/#relative-benchmarking), comparing a single run on main to a single run on a PR here: #762 This can also be accomplished with Bencher as well, and this first PR will lay the ground work for you to use [more advanced Threshold](https://bencher.dev/docs/explanation/thresholds/) for detecting performance regressions.

Hey @epompeii i saw your comment and created a [pr](https://github.com/tailcallhq/tailcall/pull/1367) i took a little bit inspiration from [3849](https://github.com/diesel-rs/diesel/pull/3849) i would be grateful if you could have a look to the pr.

**Comment by epompeii** [comment]
@tusharmath I have create a pull request to add Bencher to track the performance of the `main` branch in the `tailcall` CI: https://github.com/tailcallhq/tailcall/pull/1441

I very much appreciate the generous offer of a bounty. My goal in doing these sorts of integrations is to spread the word about Bencher. All that I ask in return is that you help spread the word in your own way, however large or small 😃

**Comment by epompeii** [comment]
I'm aware of the failure occurring here: https://github.com/tailcallhq/tailcall/actions/runs/8332008049/job/22800159465#step:5:1
And I'm investigating.

**Comment by epompeii** [comment]
@tusharmath  I'm sorry for the delay getting back to this and my failure to full test https://github.com/tailcallhq/tailcall/pull/1441, that is totally on me.

The reason for the delay is that I was heads down working on [Bencher v0.4.5](https://bencher.dev/docs/reference/changelog/#v045). This new version completely refactors the way that [branch selection](https://bencher.dev/docs/explanation/branch-selection/) is handled and it deprecates some of the options that I would have used here. I wanted to finish that work before proceeding so I didn't add options that were going to be deprecated a few days later.
Plus, the new way of doing things is much more elegant!

With that said, it looks like @alankritdabral has done a pretty nice job adding Bencher in https://github.com/tailcallhq/tailcall/pull/1679. Though he is using the now deprecated options/way of doing things.
I have left a review on the PR for the things that need to be updated: https://github.com/tailcallhq/tailcall/pull/1679#discussion_r1562545611
If he is willing to update things, then using his PR is alright with me. Otherwise, I will do my best to get this knocked out 😃

**Comment by tusharmath** [comment]
Thanks! Would love to see this integrated into Tailcall soon.

**Comment by alankritdabral** [comment]
> @tusharmath I'm sorry for the delay getting back to this and my failure to fully test #1441, that is totally on me.
> 
> The reason for the delay is that I was heads down working on [Bencher v0.4.5](https://bencher.dev/docs/reference/changelog/#v045). This new version completely refactors the way that [branch selection](https://bencher.dev/docs/explanation/branch-selection/) is handled and it deprecates some of the options that I would have used here. I wanted to finish that work before proceeding so I didn't add options that were going to be deprecated a few days later. Plus, the new way of doing things is much more elegant!
> 
> With that said, it looks like @alankritdabral has done a pretty nice job adding Bencher in #1679. Though he is using the now deprecated options/way of doing things. I have left a review on the PR for the things that need to be updated: [#1679 (comment)](https://github.com/tailcallhq/tailcall/pull/1679#discussion_r1562545611) If he is willing to update things, then using his PR is alright with me. Otherwise, I will do my best to get this knocked out 😃

Thankyou for your review means a lot :heart: . Almost done with the changes!

**Comment by alankritdabral** [comment]
@tusharmath @epompeii done with the changes. #1679 is good to go.

**Comment by epompeii** [comment]
Okay, lets finally get this knocked out!
@tusharmath I have created a new pull request: https://github.com/tailcallhq/tailcall/pull/1725
Again, my apologies for the hubris to not fully test things out the first time around. 🤦🏽‍♂️

Unfortunately, after taking a closer look at @alankritdabral PR https://github.com/tailcallhq/tailcall/pull/1679, I noticed some fundamental security vulnerabilities in his approach for handling PRs from forks. Bencher has built in defenses that probably would have prevented you all from getting pwn requested, but it would have broken your build (again). So I just went ahead and got things implemented as recommended in the docs: https://bencher.dev/docs/how-to/github-actions/#pull-requests-from-forks
